feat #2 (admin-authentification) : creation of admin authentification

nicolasrll opened the pull request.

nicolasrll committed:
feat #2 (admin-authentification) : creation of admin authentification

nicolasrll committed:
feat #2 : installation of PHPStan and corrections made in my files (after PHPStan and others)

nicolasrll committed:
feat #2 (admin-authentification) : correction on admin authentification, testing to retrieve user list

fmoy commented on config/conf.php:
```diff
@@ -0,0 +1,7 @@
+    define('DBNAME', 'blog');
+    define('DBUSER', 'nicolas');
+    define('DBPASSWORD', 'root');
+    //define('DB', '');
```
Commentaire inutile

fmoy replied:
Retirer le composer.phar de git

fmoy commented on core/AbstractManager.php:
```diff
@@ -2,7 +2,7 @@
 namespace Core;
 
-use Core\PdoConnect;
+//use Core\PdoConnect;
```
Commentaire inutile

fmoy commented on index.php:
```diff
@@ -0,0 +1,22 @@
+use Core\Dispatcher;
+use Core\AbstractManager;
+
+//error_reporting(E_ERROR | E_WARNING | E_PARSE | E_NOTICE);
```
Commentaire inutile

fmoy commented on index.php:
```diff
@@ -0,0 +1,22 @@
+    $dispatcher->dispatch();
+} catch (Exception $e) {
+    echo 'Exception reçue : ' . $e->getMessage();
+} catch (TypeError $e) {
```
Différence peut visible entre les deux catch

nicolasrll replied:
ok je supprime le TypeError

fmoy commented on src/Controller/Admin/AuthentificationController.php:
```diff
@@ -0,0 +1,56 @@
+    {
+
+        if ($this->isSubmited('authentification')) {
+            $identifiers = ($this->getFormValues('authentification'));
```
parenthèse inutile

fmoy commented on src/Controller/Admin/AuthentificationController.php:
```diff
@@ -0,0 +1,56 @@
+
+    public function loginAction()
+    {
+
```
saut de ligne inutile à moins d'y mettre un commentaire

nicolasrll replied:
Si on parle bien de la ligne 25 je viens de vérifier dans la PSR 12 et il y a bien un saut de ligne entre 2 fonctions

nicolasrll replied:
Je viens de voir à l'intérieur de la fonction..

fmoy commented on src/Controller/Admin/AuthentificationController.php:
```diff
@@ -0,0 +1,56 @@
+    {
+
+        if ($this->isSubmited('authentification')) {
+            $identifiers = ($this->getFormValues('authentification'));
```
techniquement il pourrait il y avoir plus que login / password comme un token csrf donc si tu nommes la fonction getFormValues restons logique $formValues au lieu de $identifiers

nicolasrll replied:
Le CSRF c'est mon issue #3. Ok je renome ma variable

fmoy commented on src/Controller/Admin/AuthentificationController.php:
```diff
@@ -0,0 +1,56 @@
+
+            $result = (new AdminManager())->findOne(['login' => $identifiers['login']]);
+
+            if(empty($result)){
```
Je sais pas si c'est logique de déclencher une exception, tu as déjà vu une page 404 lorsque tu essayes de t'authentifier ?ça marche pas car tu t'es trompé avec le login / password ok, mais de la a dire erreur serveur ou d'applicatif, c'est une erreur humaine donc un message en rouge dans le formulaire c est mieux.

nicolasrll replied:
non c'est pas cohérent

fmoy commented on src/Controller/Admin/AuthentificationController.php:
```diff
@@ -0,0 +1,56 @@
+    }
+
+    public function logoutAction() {
+        $_SESSION = array();
```
mouais c'est disons radical et potentiellement dangereux, peut être devrais tu utiliser une fonction qui vide ou détruit tout simplement ta session au lieu d'écraser le contenu qu'elle contient.

nicolasrll replied:
pourquoi est-ce dangereux ? 
Je voulu suivre cette documentation : https://www.php.net/manual/fr/function.session-destroy.php
* nettoyer le tableau $_SESSION
* activer session.use_strict_mode

fmoy commented on src/Controller/Admin/AuthentificationController.php:
```diff
@@ -0,0 +1,56 @@
+
+    public function logoutAction() {
+        $_SESSION = array();
+        header('Location: /admin/home/');
```
[header](https://www.php.net/manual/fr/language.basic-syntax.comments.php)

nicolasrll replied:
Je comprend pas ton commentaire. Mais si ça rejoint l'autre où tu me demandes d'ajouter exit c'est corrigé

fmoy commented on src/Controller/Admin/HomeController.php:
```diff
@@ -0,0 +1,24 @@
+            );
+        }
+
+        header('Location: /admin/authentification/login');
```
[header](https://www.php.net/manual/fr/language.basic-syntax.comments.php)

fmoy commented on src/Controller/Admin/HomeController.php:
```diff
@@ -0,0 +1,24 @@
+{
+    public function indexAction()
+    {
+        if ($_SESSION['login']) {
```
j'utiliserai à la place une fonction isAuthentified() ou quelques chose comme ça, tu ne vas pas écrire x fois cette ligne, si demain elle change ...

nicolasrll replied:
Dans mon dernier commit je l'ai retiré. Je prend en compte pour le AdminControllerAbstract

fmoy commented on src/Controller/Admin/HomeController.php:
```diff
@@ -0,0 +1,24 @@
+            return $this->renderView(
+                'back/home.html.twig',
+                [
+                    'login' => $_SESSION['login']
```
encore une fois c'est long a écrire et puis utilise le system twig a fond
`$twig->addGlobal('session', $_SESSION);`

fmoy commented on src/Entity/User.php:
```diff
@@ -0,0 +1,79 @@
+    protected $role = 'user';
+    protected $status = 'active';
+
+    public function getId(): int
```
getId ? elle existe pas dans Abstract ?

fmoy commented on src/Entity/User.php:
```diff
@@ -0,0 +1,79 @@
+{
+    protected $login = '';
+    protected $password = '';
+    protected $role = 'user';
```
je vois pas l’intérêt de mettre user si tu gères que l'administration ?

fmoy replied:
de plus fait une constante pour 'admin'

nicolasrll replied:
Pour commenter un projet je pars du principe qu'il faut être authentifié

fmoy commented on src/Repository/AdminManager.php:
```diff
@@ -0,0 +1,24 @@
+
+namespace App\Repository;
+
+//use Core\PdoConnect;
```
commentaire inutile

fmoy commented on src/Repository/AdminManager.php:
```diff
@@ -0,0 +1,24 @@
+use Core\AbstractManager;
+use App\Entity\Admin;
+
+class AdminManager extends AbstractManager
```
C'est pas logique d'avoir AdminManager si tu as une table user ???? commentaireManager = table commentaire

fmoy commented on template/back/_footer.html:
```diff
@@ -0,0 +1,7 @@
+<footer>
+    <ul>
+        <li><a href="/authentification/logout">Déconnexion</a></li>
+        <li><a href="/">Revenir au site</a></li>
```
Si tu es authentifié peut être faut il mettre /admin/

nicolasrll replied:
Revenir au site permet revenir sur le front office

fmoy commented on template/back/_footer.html:
```diff
@@ -0,0 +1,7 @@
+        <li><a href="/authentification/logout">Déconnexion</a></li>
+        <li><a href="/">Revenir au site</a></li>
+    </ul>
+    © - 2019 - Nicolas Rellier - Site de demo
```
On est en 2020 au pire met une date courante comme sa tu la changes pas chaque année

nicolasrll replied:
J'ai replacé pour utiliser la fonction date() en js que j'ai placé dans un span

fmoy commented on template/back/_header.html:
```diff
@@ -0,0 +1,12 @@
+            <li><a href="/admin/article/new">Ajouter un article</a></li>
+            <li><a href="/admin/article">Listes des articles</a></li>
+            <li><a href="/comment/list">Listes des commentaire</a></li>
+            <li><a href="#">Déconnexion</a></li>
```
je vois pas comment il marche ton lien de déconnexion ?

nicolasrll replied:
Je l'avais mis que dns le footer. C'est rectifié

fmoy commented on template/back/authentification-admin.html.twig:
```diff
@@ -0,0 +1,18 @@
+{% extends "back/template.html.twig" %}
```
je suis pas sur que la page d'authentification soit coté administration, je trouve que c est du front comme ton controleur est front lui aussi CQFD

nicolasrll replied:
Oui j'ai rectifié.
J'en ai profité pour retirer le header et le footer : je trouve ça plus sécurisé d'avoir une connexion admin sans afficher le menu

fmoy commented on template/back/template.html.twig:
```diff
@@ -0,0 +1,16 @@
+        {% include 'back/_header.html' %}
+        <main>
+            {% block content %}{% endblock %}
+            {% include 'back/_footer.html' %}
```
le footer va pas dans <footer> ?

fmoy commented on core/AdminControllerAbstract.php:
```diff
@@ -0,0 +1,27 @@
+
+namespace Core;
+
+//use Twig\Loader\FilesystemLoader;
```
Commentaire inutile

fmoy commented on core/AdminControllerAbstract.php:
```diff
@@ -0,0 +1,27 @@
+{
+    public function __construct()
+    {
+        if (!isset($_SESSION['login'])) {
```
je vois pas l’intérêt de mettre à vide une variable de session. tu devrais avoir une fonction à la limite

fmoy commented on core/AdminControllerAbstract.php:
```diff
@@ -0,0 +1,27 @@
+        }
+
+        if (false === $this->isLogged()) {
+            header('Location: /authentification/login');
```
exit apres un header.

fmoy commented on core/AdminControllerAbstract.php:
```diff
@@ -0,0 +1,27 @@
+
+    public function isLogged(): bool
+    {
+        return $_SESSION['login'];
```
$_SESSION['login'] ?? '';

fmoy commented on core/Dispatcher.php:
```diff
@@ -65,9 +65,10 @@ public function getController(): DefaultControllerAbstract
+    public function setController(string $controllerName)
     {
-        $controller = 'App\Controller\\'.$controller;
+        $controller =$this->router->isAdmin() ? 'App\Controller\Admin\\' : 'App\Controller\\';
```
$controllerFolderPath

fmoy replied:
à un new on met tjr des ()

fmoy commented on sql/blog.sql:
```diff
@@ -0,0 +1,65 @@
+--
+
+INSERT INTO `user` (`id`, `login`, `password`, `role`, `status`) VALUES
+(1, 'admin-p5', '$2y$10$06AxFg/qa2H5/lyRN/4RDOX7t17oVrI7/MOD/EUgKPq/fFLc/pGfq', 'admin', 'active');
```
peut être pourrais tu utiliser un enum pour le role ?

nicolasrll replied:
j'ai oublié de joindre le ficher sql à jour dans mon dernier push...

fmoy commented on src/Controller/Admin/HomeController.php:
```diff
@@ -1,24 +1,18 @@
+        return $this->renderView(
+            'back/home.html.twig',
+            [
+                'login' => $_SESSION['login']
```
pareil en global dans twig

fmoy commented on src/Controller/Admin/UserController.php:
```diff
@@ -0,0 +1,31 @@
+        return $this->renderView(
+            'back/home.html.twig',
+            [
+                'login' => $_SESSION['login']
```
pareil en global dans twig

fmoy commented on src/Controller/Admin/UserController.php:
```diff
@@ -0,0 +1,31 @@
+        return $this->renderView(
+            'back/users.html.twig',
+            [
+                'users' => $users
```
`'users' => (new UserManager())->find()` directement

fmoy commented on src/Controller/AuthentificationController.php:
```diff
@@ -2,14 +2,12 @@
-session_start();
-
-require_once('src/Repository/AdminManager.php');
+//require_once('src/Repository/UserManager.php');
```
commentaire inutile

fmoy commented on template/back/users.html.twig:
```diff
@@ -0,0 +1,14 @@
+    {% for user in users %}
+        <p>
+            {{ user.login }}<br>
+            {{ user.password }}<br>
```
je vois pas bien l’intérêt du password encodé mais bon

nicolasrll replied:
Finalement je l'affiche pas par sécurité.
En plus j'affiche l'id de l'utilisateur

fmoy reviewed (changes requested):
Good luck

nicolasrll committed:
feat #2 : correction made for the review

nicolasrll committed:
feat #2 : refactoring the loginAction method, error messages improvement and input form become required now

nicolasrll committed:
feat #2 : sql file updated with role as enum

fmoy commented on src/Entity/User.php:
```diff
@@ -0,0 +1,67 @@
+     */
+    public function setStatus(string $status)
+    {
+        $this->status = $status;
```
ajouter un in_array pour vérifier que le rôle transmis est autorisé

nicolasrll committed:
feat #2 : checking the role in authentification function and valued passed in setStatus and setRole and correction made in logout function

nicolasrll committed:
feat #2 : integration of bootstrap, back office authentification, home and user list done

nicolasrll committed:
feat #2 : line break added for presentation and  lunch new code climate analyse

nicolasrll committed:
feat #2 : refactorization of function authentificationChecks to have less lines and pass Code Climate. Asset folder removed from index, styles.css and conf.php are modified

fmoy commented on composer.json:
```diff
@@ -0,0 +1,21 @@
+        "phpstan/phpstan": "^0.12.20"
+    },
+    "scripts": {
+        "post-update-cmd": [
```
install

fmoy commented on config/conf.php:
```diff
@@ -0,0 +1,6 @@
+
+    define('DSN', 'mysql:host=localhost');
+    define('DBNAME', 'blog');
+    define('DBUSER', 'nicolas');
```
ne pas mettre ton vrai code normalement il devrait être ignorer genre config/conf.php.sample

fmoy commented on core/AdminControllerAbstract.php:
```diff
@@ -0,0 +1,22 @@
+        }
+    }
+
+    public function isLogged(): bool
```
chaîne vide différente de boolean

fmoy commented on core/DefaultControllerAbstract.php:
```diff
@@ -17,6 +17,7 @@ protected function renderView(string $view, array $params = [])
         $loader = new FilesystemLoader('template/');
         $twig = new Environment($loader);
         $twig->addExtension(new \Twig\Extension\DebugExtension());
+        $twig->addGlobal('session', $_SESSION);
```
ajout d'un commentaire pour la compréhension

fmoy commented on template/authentification-admin.html.twig:
```diff
@@ -0,0 +1,39 @@
+        <meta name="viewport" content="width=device-width, initial-scale=1.0">
+        <title>Espace administration - Site web personnel de Nicolas Rellier</title>
+
+        <link href="../public/bootstrap/css/bootstrap.min.css" rel="stylesheet">
```
essayer en absolute

fmoy replied:
Alors le absolute ?

nicolasrll committed:
feat #2 : enhanced templates and corrections made following the last meeting

issue#2

nicolasrll committed:
feat #2 : PHP 7.4.6 upgrading and validation of PHPStan level 8 (typed properties and functions)

issue #2

fmoy commented on core/AbstractEntity.php:
```diff
@@ -21,7 +23,7 @@ public function getId(): int
     }
 
-    public function setId(int $id)
+    public function setId(int $id): self
```
faux

fmoy commented on config/conf.php:
```diff
@@ -0,0 +1,6 @@
+<?php
+
+    define('DSN' /*, mysql:host= hote a renseigner*/);
```
```php
<?php
    // To be modified by your own configuration
    define('DSN' , 'mysql:host=localhost');
    define('DBNAME', 'blog');
    define('DBUSER', 'root');
    define('DBPASSWORD', '');
```

fmoy commented on core/AdminControllerAbstract.php:
```diff
@@ -0,0 +1,22 @@
+
+    public function isLogged(): bool
+    {
+        return $_SESSION['login'] ?? false;
```
C'est plus sur de tester la valeur booleen "true", peut être peux tu trouver un autre nom que login, car c est souvent l'email ou le login de la personne.
```php
<?php
return isset($_SESSION['isLogged']) 
    && true === $_SESSION['isLogged'];
```

nicolasrll replied:
justement $_SESSION['login'] contient le login de l''admin que j'affiche sur la home

fmoy commented on core/DefaultControllerAbstract.php:
```diff
@@ -17,6 +17,7 @@ protected function renderView(string $view, array $params = [])
         $loader = new FilesystemLoader('template/');
         $twig = new Environment($loader);
         $twig->addExtension(new \Twig\Extension\DebugExtension());
+        $twig->addGlobal('session', $_SESSION); // share php session variable
```
Un commentaire c est en haut pas sur le coté

fmoy commented on public/css/styles.css:
```diff
@@ -0,0 +1,205 @@
+@charset "UTF-8";
```
il est versionné ce fichier ?

nicolasrll replied:
C'est le fichier style de mon template après l'avoir nettoyé (il faisait 10 000 lignes) alors oui

fmoy commented on public/js/scripts.js:
```diff
@@ -0,0 +1,56 @@
+/*!
```
il est versionné ce fichier ?

nicolasrll replied:
bah oui je comprend pas

fmoy commented on src/Controller/Admin/HomeController.php:
```diff
@@ -0,0 +1,15 @@
+
+class HomeController extends AdminControllerAbstract
+{
+    public function indexAction(): AdminControllerAbstract
```
euh ... nan, ca retourne pas $this, mais $this->renderView();

nicolasrll replied:
Je sais. AdminControllerAbstract hérite de DefaultControllerAbstract.
D'ailleurs j'ai essayé avec DefauktControllerAbstract et PHPStan me renvoit une erreur

fmoy commented on src/Controller/Admin/UserController.php:
```diff
@@ -0,0 +1,19 @@
+
+class UserController extends AdminControllerAbstract
+{
+    public function indexAction(): AdminControllerAbstract
```
euh ... nan, ca retourne pas $this, mais $this->renderView();

fmoy commented on src/Controller/AuthentificationController.php:
```diff
@@ -0,0 +1,75 @@
+
+class AuthentificationController extends DefaultControllerAbstract
+{
+    public function indexAction(): DefaultControllerAbstract
```
euh ... nan, ca retourne pas $this, mais $this->renderView();

fmoy commented on src/Controller/HomeController.php:
```diff
@@ -0,0 +1,19 @@
+
+class HomeController extends DefaultControllerAbstract
+{
+    public function indexAction(): DefaultControllerAbstract
```
euh ... nan, ca retourne pas $this, mais $this->renderView();

fmoy commented on src/Entity/User.php:
```diff
@@ -0,0 +1,76 @@
+    protected string $password = '';
+    protected string $role = self::USER;
+    protected bool $isActive = false;
+    protected const ADMIN = 'admin';
```
constante en haut, techniquement t'appel une constant self::user qui est définie que quelques lignes plus tard, par principe

fmoy commented on src/Entity/User.php:
```diff
@@ -0,0 +1,76 @@
+        $existingRole = [self::ADMIN, self::USER];
+
+        if (!in_array($role, $existingRole)) {
+            throw new Exception('La valeur passé n\'est pas valide');
```
Le role $role n'existe pas ou n'est pas autorisé. Essaye d'être plus précis sur des exceptions, sinon tu vas pas savoir quelle est la valeur ni même que tu étais en train de passer un rôle.

fmoy commented on src/Entity/User.php:
```diff
@@ -0,0 +1,76 @@
+        return $this;
+    }
+
+    public function getisActive(): bool
```
tu as le droit d'utiliser par exemple si tu avais une propriété enabled le getter isEnabled() au lieu de getEnabled()

fmoy commented on template/layout.html.twig:
```diff
@@ -0,0 +1,27 @@
+        <link href="/public/bootstrap/css/bootstrap.min.css" rel="stylesheet">
+        <link href="/public/css/styles.css" rel="stylesheet" />
+        <script src="https://use.fontawesome.com/releases/v5.12.1/js/all.js" crossorigin="anonymous"></script>
+        <title>{% block title %}{% endblock %}</title>
```
A voir si il en faut un par défaut ? au cas ou tu oublies de le renseigner sur une vue enfant ?

nicolasrll committed:
feat #2 : modifications following the last remarks on PR and templates clean up

issue #2

nicolasrll committed:
feat #2 : Correction made following the last meeting.

Function roleCheck moved from AuthentificationController to AdminControllerAbstract.
Corrections made on return type of some function
Creation of the databas added in sql file

issue #2

nicolasrll committed:
feat #2 : refactoring of the indexAction fuction and deletion of loginAction

nicolasrll committed:
feat #2 : correction and core made on issue 3 by mistake

Composer.json modified so as not to version the content of public folder
AdminControllerAbstract modified so as not have error on page where we're not connected
AuthentificationController refactorized

issue #2

nicolasrll committed:
feat #2 : .gitignored is now working on public folder and .composer.json modified on public folder

issue #2

nicolasrll committed:
feat #2 : correction made on path to bootstrap in layout

issue #2
